ath79: add support for Comfast CF-EW72

macromorgan opened the pull request:
Specifications:
Qualcomm/Atheros QCA9531 + QCA9886
2x 10/100 Mbps Ethernet, with 48v PoE
2T2R 2.4 GHz, 802.11b/g/n
2T2R 5 GHz, 802.11a/n/ac
128MB RAM
16MB SPI Flash
4x LED (Always On Power, LAN, WAN, WLAN)

Flashing Instructions:
Original firmware is based on OpenWRT, so flashing the sysupgrade image on the factory firmware is sufficient.

Tested:  Reset button, WAN LED, LAN LED, Power LED (always on, not much to test), WLAN LED (one LED only for 2 interfaces, by default it gets assigned to the first interface), MAC addresses (match factory firmware).

dmesg log (this pull request on master): https://gist.github.com/macromorgan/93aa428b659b1bdeb90db6fb9fd9dfc2

adapted from this forum post by rpc: https://forum.openwrt.org/t/solved-support-for-comfast-cf-ew72/29978/8

original work based on this pull request from Comfast: https://github.com/fall513/openwrt/commit/1e1e29e5d62dcd48833e71ac462e1975df7cfcf6

adrianschmutzler commented:
Your PR has formal issues. Please have another look at https://openwrt.org/submitting-patches

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins>;
+
+		lan: lan {
```
Should be
`led_lan: lan {`

Please update below accordingly.

However, at the moment it does not look like you'd be using the lan label at all? In that case, remove it.

Which LED is used by the stock FW for signalling boot etc.?

macromorgan replied:
I'm honestly not sure, I'm not on the stock firmware anymore and that's one thing I didn't check prior to getting onto this version of OpenWRT.  To my knowledge it's the WAN.

macromorgan replied:
Confirmed on the factory firmware it is the WAN LED used for signaling purposes.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+
+	keys {
+		compatible = "gpio-keys"; // -polled
+		// poll-interval = <20>;
```
remove comments

macromorgan replied:
Acknowledged.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
```
Add empty line after status.

macromorgan replied:
Acknowledged.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor" ;
```
any reason for the space before the semicolon?

macromorgan replied:
Nope, removed.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+				read-only;
+			};
+
+			firmware: partition@20000 {
```
it looks to me as if only the art node label is used. remove the rest (uboot:, firmware:, nvram:).

macromorgan replied:
Acknowledged.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+
+&eth0 {
+	status = "okay";
+	phy-handle = <&swphy4>;
```
Add empty line after status.

macromorgan replied:
Acknowledged.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,129 @@
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x6>;
```
That's unprobable. Can you please provide a list of the MAC addresses on flash (art 0x0, 0x6, 0x1002, ...) and a list of the addresses which are used on stock FW (lan, wan, 2.4 GHz, ...)

For further reading: https://openwrt.org/docs/guide-developer/mac.address

macromorgan replied:
20:0d:b0:75:9b:f2 (0x0)
20:0d:b0:75:9b:f4 (0x6)
20:0d:b0:75:9b:f3 (0x1002)

These values match the stock firmware.  Do you need a dump of the ART partition?

adrianschmutzler replied:
Well, it looks strange to me that you would "overwrite" the MAC address in the WiFi's calibration data (0x1002) with another one (0x6), but then use the WiFi address on eth1 ...

macromorgan replied:
Should I just remove the mtd-mac-address = <&art 0x6>; line?

adrianschmutzler replied:
That depends on which of those the two MAC addresses matches the 2.4G's address on stock firmware...

macromorgan replied:
I'll attempt to reflash factory tonight and confirm.

macromorgan replied:
Confirmed the LAN port is F2, the WAN port is F3, the 5G radio is F4, and the 2.4G radio is F5 on the factory firmware.

macromorgan replied:
And in this firmware it looks like the 5G is F5 and the 2.4G is F4... weird.

Is there a way I can ensure the order is swapped, or does it matter for the WiFi interface (the WAN and LAN are correct)?  Sorry, still new to all this.

adrianschmutzler replied:
So, let's collect our data:

```
use   stock    location
lan    f2         0x0
wan  f3          0x1002
5g     f4          0x6
2g     f5          0x5006
```

So, unless you mixed something up, I suspect the vendor just doesn't care what's stored in the calibration data and just increments the MAC address for its devices.

An based on this mess (I hesitate to use 5G calibration data to assign 2.4 G address, I'd recommend to just calculate all addresses based on an offset from 0x0, so

lan 0x0
wan 0x0 +1
...

adrianschmutzler replied:
Does the device have a label with a MAC address on it?

macromorgan replied:
Yes, but only for the LAN port.  There is no label for WAN or for the 2G/5G radios.

adrianschmutzler replied:
Okay, that's what I expected.

macromorgan replied:
So should I just define the LAN as-is, WAN as lan +1, and use the calibration data (remove the mtd-mac-address) line?

adrianschmutzler replied:
And please add a nicer version of my improvised table to the commit message after you have squashed your changes.

adrianschmutzler replied:
no, the calibration data would yield the wrong addresses.

macromorgan replied:
Okay.  eth0 in the devicetree (wan port, eth1 in openwrt):
        phy-handle = <&swphy4>;
        mtd-mac-address = <&art 0x0>;
        mtd-mac-address-increment = <1>;

eth1 in the devicetree (lan port, eth0 in openwrt):
        mtd-mac-address = <&art 0x0>;

wmac in the devicetree (2.4g in openwrt):
        mtd-mac-address = <&art 0x0>;
        mtd-mac-address-increment = <3>;

And coalescing cf-ew72 into the same 11-ath10k-caldata section as cf-e560ac gives me the exact same MAC addresses on Openwrt as they do on stock.  Now just to figure out the LEDs...

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -80,6 +80,11 @@ comfast,cf-e560ac)
+comfast,cf-ew72)
+	ucidef_set_led_netdev "lan" "LAN" "$boardname:blue:lan" "eth0"
+	ucidef_set_led_netdev "wan" "WAN" "$boardname:blue:wan" "eth1"
+	ucidef_set_led_wlan "wlan" "WLAN" "$boardname:blue:wlan" "phy0tpt"
```
The last line is already set in DTS.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -80,6 +80,11 @@ comfast,cf-e560ac)
 	;;
+comfast,cf-ew72)
+	ucidef_set_led_netdev "lan" "LAN" "$boardname:blue:lan" "eth0"
+	ucidef_set_led_netdev "wan" "WAN" "$boardname:blue:wan" "eth1"
```
You will have to use ucidef_set_led_switch for one of the interface to properly indicate connection state.

macromorgan replied:
I'm afraid I don't understand.  Something like the following:
ucidef_set_led_switch "lan" "LAN" "$boardname:blue:lan" "switch0" "0x02"

Sorry, this is my first new device so I'm still new to devicetree and LED configurations.

adrianschmutzler replied:
lan and wan port are connected differently internally, but so only one can properly show connection state with ucidef_set_led_netdev. The other one will need something like you have provided, only with the correct value for the last parameter...

macromorgan replied:
Where would "switch0" and "0x02" be defined so I know what values to use?

macromorgan replied:
I just confirmed there doesn't appear to be any internal switches on this device.  Additionally, this current configuration works... the LAN port is lit up when the cable is plugged in and blinks with activity, when unplugged the LED turns off (both the LAN and WAN port support PoE).  When the WAN port is plugged in it turns on and blinks with activity, and when unplugged it turns off.  When both plugged in both work (activity) independent of each other.  Unfortunately it's impossible to test with both unplugged.

Additionally, it seems this is how the Comfast CF-E314N-v2 is configured in the same LED script.

I'm going to mark this as "resolved", but if I do need to revisit this please let me know.  Thank you.

adrianschmutzler replied:
Both comfast,cf-e313ac and comfast,cf-e5 share your setup, and they use `ucidef_set_led_switch "lan" "LAN" "$boardname:blue:lan" "switch0" "0x02"`. If I have time tomorrow, I'll try to look for the explanation ...

macromorgan replied:
Okay, I'll try that and see what happens...

macromorgan replied:
Changing:
ucidef_set_led_netdev "lan" "LAN" "$boardname:blue:lan" "eth0"
To:
ucidef_set_led_switch "lan" "LAN" "$boardname:blue:lan" "switch0" "0x02"
also seems to work.  The blink pattern is noticably different, but still works more or less as expected (blinks with activity, but more rhythmic/slower).  I think the behavior with the netdev line is preferable, but since this works too I'm okay with it if it conforms to the code guidelines.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -146,6 +146,12 @@ case "$FIRMWARE" in
+	comfast,cf-ew72)
+		caldata_extract "art" 0x5000 0x2f20
+		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
+		/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
```
Please use proper indent here. Does the interface get a proper MAC address?

macromorgan replied:
Will do.  MAC addresses are 20:0D:B0:75:9B:F4 for the 2.4G and 20:0D:B0:75:9B:F5 for the 5G.

macromorgan replied:
Okay, so the one strange thing is that the MAC addresses are backwards (stock F4 is for the 5G, for this it looks like F5 is for the 5G though and F4 is for the 2.4G).  Any advice to fix?

macromorgan commented:
I just noticed I did a pull request for a series of commits rather than a single one.  Should I close this (after I fix the issues) and resubmit as a single commit?

adrianschmutzler commented:
No, please just squash your changes into a single commit (git rebase -i ...) and then force-push (git push -f) to this branch. Please do _not_ open a new PR for the same topic, as we will lose all history then.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -146,6 +146,12 @@ case "$FIRMWARE" in
 		rm /lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin
 		;;
+	comfast,cf-ew72)
+		caldata_extract "art" 0x5000 0x2f20
```
You would use
ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +2)
here after our new information, and thus can just merge this case with the previous.

macromorgan replied:
Okay, I think I get it now.  I'm just going to collapse with the above then since now it's identical to the CF-E560AC.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,132 @@
+	status = "okay";
+
+	phy-handle = <&swphy4>;
+	mtd-mac-address = <&art 0x0>;
```
This looks like WAN to me, so use
```
mtd-mac-address = <&art 0x0>;
mtd-mac-address-increment = <1>;
```

LAN will use
```
mtd-mac-address = <&art 0x0>;
```

and wmac will use
```
mtd-mac-address = <&art 0x0>;
mtd-mac-address-increment = <3>;
```

macromorgan replied:
Okay, I understand, but eth0 is actually lan and eth1 is wan.  wmac is the 2.4G, and the caldata MAC is the 5G.  I'll test all these changes and make sure the MACs are all matching the stock firmware.  I can confirm with the latest commits the LEDs all mimic stock behavior, including the WAN LED being the one that blinks for power/flash/control/etc.

Now if I can just figure out how to squash my changes properly...

adrianschmutzler replied:
Note that eth0 on the running device might be different from &eth0 in the device tree. Having `phy-handle = <&swphy4>;` indicates to me that this eth is probably wan.

macromorgan replied:
Okay, I'll test it either way and if it's wrong switch it.

macromorgan replied:
You were right...

macromorgan committed:
ath79: add support for Comfast CF-EW72

Specifications:
Qualcomm/Atheros QCA9531 + QCA9886
2x 10/100 Mbps Ethernet, with 48v PoE
2T2R 2.4 GHz, 802.11b/g/n
2T2R 5 GHz, 802.11a/n/ac
128MB RAM
16MB SPI Flash
4x LED (Always On Power, LAN, WAN, WLAN)

Flashing Instructions:
Original firmware is based on OpenWRT, so flashing the sysupgrade image on
the factory firmware is sufficient.

Tested: Reset button, WAN LED, LAN LED, Power LED (always on, not much
to test), WLAN LED (one LED only for 2 interfaces, by default it gets
assigned to the first interface), MAC addresses (match factory firmware).
My LAN factory MAC address ends in F2.

use	stock_mac	art_loc
lan	:f2		0x0
wan	:f3		0x1002
5g	:f4		0x6
2g	:f5		0x5006

Signed-off-by: Chris Morgan <macromorgan@hotmail.com>

macromorgan commented:
I believe I've addressed all of the concerns with this pull request; if I am mistaken or there is anything else that I need to address please let me know.  Please double check my comments on the 01_leds file as I dissented with your comment (which 95% of the time means I'm wrong when I disagree with an expert, but who knows if this is a 5%).

Thank you for all your help.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,134 @@
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		wan: wan {
```
`led_wan: wan {`
Update aliases accordingly.

macromorgan replied:
Whoops, that got reverted when I fumbled the first few squash attempts.  Sorry.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts:
```diff
@@ -0,0 +1,134 @@
+};
+
+&eth1 {
+	status = "okay";
```
status can be removed here (eth1), it is already set in qca953x.dtsi

macromorgan replied:
Acknowledged.

macromorgan commented:
I've made the requested changes to the 01_leds file.  Let me know if there is anything else that remains to be done.  Thank you.

openwrt-diy commented:
Comfast CF-EW72  Wireless 5.8G often fakes for no reason, the signal cannot be connected! This happens three or four times a week. 
我刷的固件是https://github.com/fall513/openwrt/tree/ar71xx-add-support-for-COMFAST-CF-EW71/EW72 还有 openwrt-18.06.2，openwrt-19.07.2,

openwrt-diy commented:
Comfast CF-EW72 Wireless 5.8G often fakes for no reason, the signal cannot be connected! This happens three or four times a week.
![CF-EW72 - 总览](https://user-images.githubusercontent.com/57249132/76597516-9d9cdb80-653b-11ea-9533-c09ee3c6e50d.png)
![CF-EW72 - 无线](https://user-images.githubusercontent.com/57249132/76597524-a2fa2600-653b-11ea-8b97-bbe9422df0a0.png)
![设备配置](https://user-images.githubusercontent.com/57249132/76597533-a68dad00-653b-11ea-9a88-f1ac303791db.png)

adrianschmutzler commented:
@openwrt-diy You are using the ar71xx version. This is for supporting the device on ath79. So, the question would be whether you have the same issues with ath79 as well.

adrianschmutzler commented:
Your patch still lacks a Signed-off-by line, despite I'd remove the links at the bottom and put that into the forum, as I wouldn't expect those to be persistent enough for putting them in a git message.

And please make the "add" in the commit title lower case.

macromorgan commented:
Changes made.  And I'll keep an eye on the 5G stability, but so far I haven't noticed anything awry.

Thank you.

openwrt-diy commented:
> @openwrt-diy You are using the ar71xx version. This is for supporting the device on ath79. So, the question would be whether you have the same issues with ath79 as well.

Tomorrow, I will try to compile the ath79 support scheme you provided!
I have been using the cf-e72 outdoors for nearly three months. I want to try to avoid the interference automatically after I turn on the cf-e72.

openwrt-diy commented:
> Changes made. And I'll keep an eye on the 5G stability, but so far I haven't noticed anything awry.
> 
> Thank you.

Has successfully flashed the ATH79 firmware, currently 5.8G is working normally! Wifi speed up a lot
![下载](https://user-images.githubusercontent.com/57249132/76694462-14f27c80-66ae-11ea-88c5-03f613e42979.png)

macromorgan commented:
Any more issues I need to address before this can be mainlined?  Let me know and I'll get them fixed.

Thank you.

macromorgan commented:
Thank you very much.  I'm going to deploy my device to production now (was waiting in case I needed to test anything else like LEDs or something).

GoliathLabs commented:
@macromorgan I have an EW71 v2 and plan to add proper support for it as well. Out of curiosity: Does your EW72 also have TFTP recovery on boot? As soon as mine gets to uboot, it starts a tftp recovery from 192.168.1.10 with the firmware_auto.bin file.
